### ISSUE TITLE
[video_player] Add Windows support

### ISSUE THREAD

**Issue by Tumbler0809** [flutter/flutter#37673]
When does the video player plugin support the Windows platform? Is there any plan support?（https://pub.dev/packages/video_player）

**Comment by Tumbler0809** [comment]
Can video playback support full-screen playback?

**Comment by stuartmorgan** [comment]
Currently blocked on:

- Finalized plugin API
- Texture support for the Win32 embedding (once that lands)

**Comment by NubilaPardus** [comment]
Is there any option to open windows vlc.exe under our flutter desktop app (from dart code) like cmd(bash) command
$ ..../VideoLAN/VLC/vlc.exe "https://...."

**Comment by NubilaPardus** [comment]
This is OK for me. (also waiting windows plugin of flutter video player)

String newUrl = "http:.....";
Process.run('vlc', ['$newUrl']).then((ProcessResult results) { print(results.stdout); });

**Comment by thesmartrenovation** [comment]
> This is OK for me. (also waiting windows plugin of flutter video player)
> 
> String newUrl = "http:.....";
> Process.run('vlc', ['$newUrl']).then((ProcessResult results) { print(results.stdout); });

Does this work in a Windows Flutter APP? I mean, can I launch VLC for a specific URL?

**Comment by NubilaPardus** [comment]
> > This is OK for me. (also waiting windows plugin of flutter video player)
> > String newUrl = "http:.....";
> > Process.run('vlc', ['$newUrl']).then((ProcessResult results) { print(results.stdout); });
> 
> Does this work in a Windows Flutter APP? I mean, can I launch VLC for a specific URL?

I could launch vlc player on flutter app with using given code. I wrote this code a year ago.

**Comment by thesmartrenovation** [comment]
> > > This is OK for me. (also waiting windows plugin of flutter video player)
> > > String newUrl = "http:.....";
> > > Process.run('vlc', ['$newUrl']).then((ProcessResult results) { print(results.stdout); });
> > 
> > 
> > Does this work in a Windows Flutter APP? I mean, can I launch VLC for a specific URL?
> 
> I could launch vlc player on flutter app with using given code. I wrote this code a year ago.

I can't make it work, I want to open VLC with a specific URL, but I get this error: 

```
[ERROR:flutter/lib/ui/ui_dart_state.cc(166)] Unhandled Exception: ProcessException: The system cannot find the specified file.


  Command: vlc http:.....
```
Could you give me some guidance on how you made it work? I would really appreciate it

**Comment by NubilaPardus** [comment]
> > > > This is OK for me. (also waiting windows plugin of flutter video player)
> > > > String newUrl = "http:.....";
> > > > Process.run('vlc', ['$newUrl']).then((ProcessResult results) { print(results.stdout); });
> > > 
> > > 
> > > Does this work in a Windows Flutter APP? I mean, can I launch VLC for a specific URL?
> > 
> > 
> > I could launch vlc player on flutter app with using given code. I wrote this code a year ago.
> 
> I can't make it work, I want to open VLC with a specific URL, but I get this error:
> 
> ```
> [ERROR:flutter/lib/ui/ui_dart_state.cc(166)] Unhandled Exception: ProcessException: The system cannot find the specified file.
> 
> 
>   Command: vlc http:.....
> ```
> 
> Could you give me some guidance on how you made it work? I would really appreciate it

You should install vlc player on windows before using it. Then,  you should set vlc as environment variable (hint: https://www.vlchelp.com/add-vlc-command-prompt-windows/). I have not used flutter for a year, i could not add any code in here.

**Comment by thesmartrenovation** [comment]
> > > > > This is OK for me. (also waiting windows plugin of flutter video player)
> > > > > String newUrl = "http:.....";
> > > > > Process.run('vlc', ['$newUrl']).then((ProcessResult results) { print(results.stdout); });
> > > > 
> > > > 
> > > > Does this work in a Windows Flutter APP? I mean, can I launch VLC for a specific URL?
> > > 
> > > 
> > > I could launch vlc player on flutter app with using given code. I wrote this code a year ago.
> > 
> > 
> > I can't make it work, I want to open VLC with a specific URL, but I get this error:
> > ```
> > [ERROR:flutter/lib/ui/ui_dart_state.cc(166)] Unhandled Exception: ProcessException: The system cannot find the specified file.
> > 
> > 
> >   Command: vlc http:.....
> > ```
> > 
> > 
> > Could you give me some guidance on how you made it work? I would really appreciate it
> 
> You should install vlc player on windows before using it. Then, you should set vlc as environment variable (hint: https://www.vlchelp.com/add-vlc-command-prompt-windows/). I have not used flutter for a year, i could not add any code in here.

Thank you very much, I've already made it work.

**Comment by stuartmorgan** [comment]
Please don't use this issue for discussions that aren't related to implementing support for `video_player` plugin. StackOverflow, Discord, or other general forums would be the right place to discuss alternate approaches to video playback.

**Comment by Bonfra04** [comment]
Hi, it's been a while and I've explored a lot of issues but found nothing. How is the situation with the Texture implementation on windows? Are there any chance to see the video_player plugin updated soon?
Since `video_player 0.10.9+2 (May 6 2020)` in the README file is included this statement:
> Backward compatible 1.0.0 version is coming
But now we are on 0.11.0 and still nothing. Is it blocked by something or is just a long process to implement?

**Comment by leielyq** [comment]
How to use the library on  windows?Thanks

**Comment by leielyq** [comment]
Any news?sad

**Comment by Bonfra04** [comment]
@leielyq Nothing changed... As far as I can understand there was a blocking feature with Textures for windows. Textures are needed to render "custom" content to a flat surface. If this feature has been implemented in flutter for windows now it should be a really small time for the plugin to make his way into windws. I'm waiting for it too.

**Comment by syleishere** [comment]
Using simple example on plugin page I get following:

Building Windows application...
Debug service listening on ws://127.0.0.1:64992/c7h-dPiYBYM=/ws
Syncing files to device Windows...
[ERROR:flutter/lib/ui/ui_dart_state.cc(186)] Unhandled Exception: PlatformException(channel-error, Unable to establish connection on channel., null, null)
#0      VideoPlayerApi.initialize (package:video_player_platform_interface/messages.dart:177:7)
<asynchronous suspension>

[ERROR:flutter/lib/ui/ui_dart_state.cc(186)] Unhandled Exception: PlatformException(channel-error, Unable to establish connection on channel., null, null)
#0      VideoPlayerApi.create (package:video_player_platform_interface/messages.dart:199:7)
<asynchronous suspension>
#1      MethodChannelVideoPlayer.create (package:video_player_platform_interface/method_channel_video_player.dart:46:31)
<asynchronous suspension>
#2      VideoPlayerController.initialize (package:video_player/video_player.dart:282:18)
<asynchronous suspension>

**Comment by yangyxd** [comment]
There is a dart_vlc  plug-in in the pub repository, I have used to achieve video playback in windows platform, the effect is great.

**Comment by unoravi** [comment]
@yangyxd 
I am facing few issues on dart_vlc plig-in : https://github.com/alexmercerind/dart_vlc/issues/59

Please let me know if any.

Thanks

**Comment by unoravi** [comment]
@all
Can anyone share approximate date/month for this development ?

Thanks

**Comment by raitonoberu** [comment]
If only it were that simple.
The fact is that android and iOS have built-in APIs for playing media. Windows and Linux don't have such APIs, so we have to use external libraries. So far, dart_vlc is the best solution.

**Comment by anirudhb** [comment]
Hi, I've written a Windows implementation of the video_player platform interface. You can check it out here: [video_player_windows](https://github.com/anirudhb/flutter_packages). All you have to do is add it as a dependency like so (along with the normal video_player package):
```yaml
dependencies:
  video_player_windows:
    git:
      url: https://github.com/anirudhb/flutter_packages.git
      path: video_player_windows/
```
It uses FFmpeg and is still a bit rough around the edges but I have found that it works well for basic video playback. Contribution is appreciated to support playback speed so this may be able to get merged upstream :-)

**Comment by stuartmorgan** [comment]
> It uses FFmpeg (with libao for audio playback) 

Interesting, is there a reason you went with a separate library rather than something built into the OS (e.g., Media Foundation)?

> You can check it out here: [video_player_windows](https://github.com/anirudhb/flutter_packages). 

Please consider renaming it (to, e.g., video_player_ffmpeg_windows). `video_player_windows` is the name we intend to use for the official package, so it would be extremely unfortunate if someone published your package to pub. (You may not think anyone would ever do that, but as someone who maintains a deliberately unpublished package that someone decided to publish without my involvement, I can say from experience that it can happen.)

**Comment by stuartmorgan** [comment]
> so this may be able to get merged upstream

To set expectations here: the current thinking based on experience with the mobile implementations (especially around DRM) is that we may well wait for platform view support for desktop implementations of video_player. So there's a larger design question to resolve before any implementation would be merged into flutter/plugins, beyond just quality/completeness evaluation.

However, one of the significant advantages of federation is that even if the official implementation goes a completely different direction, people could choose to swap in your implementation if it better suited their use case. There doesn't have to be only one implementation for a given front-end package for each platform under the federated model.

**Comment by anirudhb** [comment]
> Interesting, is there a reason you went with a separate library rather than something built into the OS (e.g., Media Foundation)?

WMF did not seem to support offscreen rendering, so it seemed like a platform-view solution rather than using the already available texture APIs.

> Please consider renaming it (to, e.g., video_player_ffmpeg_windows). 

Will do.

**Comment by Tonku** [comment]
> > so this may be able to get merged upstream
> 
> To set expectations here: the current thinking based on experience with the mobile implementations (especially around DRM) is that we may well wait for platform view support for desktop implementations of video_player. So there's a larger design question to resolve before any implementation would be merged into flutter/plugins, beyond just quality/completeness evaluation.


What there is no platform view support yet for desktop. It been ages :(

**Comment by unoravi** [comment]
@stuartmorgan 
Please keep me posted if any updates on this..

**Comment by unoravi** [comment]
Any updates on this ??

**Comment by azchohfi** [comment]
@stuartmorgan Whichever way the Windows implementation goes, does it have a requirement of supporting a minimum version of Windows or would it be ok to implement this using Xaml Islands and using the [MediaPlayerElement](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.mediaplayerelement?view=winrt-20348) APIs? It would be Windows 10 only, but should work fine in Win32 or UWP, and it is a very fully featured control. It is also maintained by Microsoft, so it is a safe bet and also very standard, right? I would probably be able to send a PR with this. Would it be up for grabs?

**Comment by stuartmorgan** [comment]
Currently, Flutter's support for Windows versions below 10 is best-effort, so a `video_player` implementation that only supports Windows 10 would be fine (and if someone really wanted to implement a fallback for earlier versions, they could always contribute a PR for that later—or even do it as an alternate, non-endorsed implementation).

> I would probably be able to send a PR with this.

Not until there is platform view support for Windows though, unless there's something about the proposal I'm not understanding.

**Comment by azchohfi** [comment]
I've managed to implement a ~~terrible~~ version of the video player using Xaml Islands. Now I have a better, but incomplete version (only audio) using the MediaFoundation APIs. For video to work, it will require [DirectComposition support](https://github.com/flutter/engine/pull/24705), as well as a new [TextureVariant](https://github.com/flutter/engine/blob/master/shell/platform/common/client_wrapper/include/flutter/texture_registrar.h#L47-L48) for providing a Visual instead of a callback for copying pixels.`
During this process, I've also partially implemented C++ code generation for Pigeon, which I'll soon also send a PR.

**Comment by stuartmorgan** [comment]
> I've also partially implemented C++ code generation for Pigeon, which I'll soon also send a PR.

Excellent, this is something that was on the list of thing we were hoping to be able to get to relatively soon.

**Comment by azchohfi** [comment]
> > I've also partially implemented C++ code generation for Pigeon, which I'll soon also send a PR.
> 
> Excellent, this is something that was on the list of thing we were hoping to be able to get to relatively soon.

https://github.com/flutter/packages/pull/476

**Comment by alexmercerind** [comment]
> I've managed to implement a ~terrible~ version of the video player using Xaml Islands. Now I have a better, but incomplete version (only audio) using the MediaFoundation APIs. For video to work, it will require [DirectComposition support](https://github.com/flutter/engine/pull/24705), as well as a new [TextureVariant](https://github.com/flutter/engine/blob/master/shell/platform/common/client_wrapper/include/flutter/texture_registrar.h#L47-L48) for providing a Visual instead of a callback for copying pixels.` During this process, I've also partially implemented C++ code generation for Pigeon, which I'll soon also send a PR.

I tried to retrieve the pixel buffer of video frames, but I cannot create an `IDirect3DSurface` (or something that implements it). According to [this article](https://docs.microsoft.com/en-us/windows/uwp/audio-video-camera/play-audio-and-video-with-mediaplayer#use-mediaplayer-in-frame-server-mode), I used `IsVideoFrameServerEnabled` to enable the `VideoFrameAvailable` callbacks & those get invoked, but not having a surface to copy to, it just results in crash.

Next, I posted a stackoverflow question, they said me to make a DXGI surface & convert it to `IDirect3DSurface` and extract frame buffer from it. But I have no idea.

My current attempt: [Here](https://github.com/harmonoid/libwinmedia/blob/99e20c240c35f0480f7959724be2598ed1f12981/libwinmedia.cpp#L887-L913).

I'm already using it for audio playback in my app (working flawless). Direct X isn't exactly what I know about too much. XAML islands just always overlap the win32 content, not exactly pleasant.

Looking forward.

**Comment by tinywolf3** [comment]
I'm using now the dart_vlc package for windows videos.
https://pub.dev/packages/dart_vlc

**Comment by sudongg** [comment]
I want the Flutter video player on desktop to be able to play 1080 video at 60fps at least.
I've used the dart_vlc package and have noticeable frame drops when playing 1080 60fps video, more so when playing 4k video.
uwp MediaPlayerElement on Windows is a good reference, but it also suffers from frame shortages when playing 4k 60fps video.
Google Chrome can play 4k 60fps video perfectly. I wonder if the Flutter video player can get inspiration from it?

**Comment by alexmercerind** [comment]
> I want the Flutter video player on desktop to be able to play 1080 video at 60fps at least. I've used the dart_vlc package and have noticeable frame drops when playing 1080 60fps video, more so when playing 4k video. uwp MediaPlayerElement on Windows is a good reference, but it also suffers from frame shortages when playing 4k 60fps video. Google Chrome can play 4k 60fps video perfectly. I wonder if the Flutter video player can get inspiration from it?

Since I'm author of `dart_vlc`, I think I should make a reply accounting the cause of this bad performance. 
No matter what video playback library you use right now, it can't be hardware accelerated because as soon as you convert the copy video frames into array buffer (likely `uint8_t[]`) (which is then used by Flutter) it becomes something that CPU does (and not just GPU). It is really expensive process. Currently only available texture type for Windows is `PixelBufferTexture` which is never going to have a good performant (hardware accelerated) video playback. However, we have this: https://github.com/flutter/engine/pull/26840 (but will take another eternity to get merged).
Copying video frame into pixel buffer every frame isn't performant & just becomes software rendering.
Ideally most APIs like one from libVLC 4.0.0 or libMPV allow to render to an OpenGL context or a Direct3D surface which is actually hardware accelerated & performant. 

Another performant approach can be to use somthing like `SetParent` or passing Flutter window's `HWND` to libVLC or libMPV directly to render video on it, but it won't allow any Flutter widgets on top of the video viewport (I have done this) & sometimes boundaries will mismatch with Flutter's widgets.

**Comment by sudongg** [comment]
@alexmercerind 
I have tried to use LibVLCVideoSetCallbacks + SkiaSharp in LibVLC in the past, and I have turned on SkiaSharp GPU rendering, and I can't achieve smooth playback at 1080 60fps.
So I didn't expect better results before using it.

For the player implementation of the GUI framework, at present I have tested the ones that can achieve 1080 60fps smooth playback on windows:

1.wpf + D3DImage +  directshow evr cp
2.uwp + CompositeSurface + MediaFoundation
3.QT + openGL+ LibMpv
4.Google Chrome

Google Chrome has the best playback effect, uwp can meet most use cases (1080 60fps, 4k 30fps), and wpf QT basically satisfies 1080 60fps playback, without any obvious frame loss.

Flutter video player should meet the requirements of 1080 60fps, 4k 30fps, otherwise it is meaningless.
Window handle rendering does Not suitable with Flutter video player , most users want the Flutter video player to be a widget.
but if you want to make a player application, it is the best way to use the window handle to render + use SetParent to make the Flutter native window as a transparent overlay (I have used SkiaSharp to implement it in winforms, but I have not implemented it in Flutter, I think it should be To modify the engine source code).

**Comment by alexmercerind** [comment]
> @alexmercerind I have tried to use LibVLCVideoSetCallbacks + SkiaSharp in LibVLC in the past, and I have turned on SkiaSharp GPU rendering, and I can't achieve smooth playback at 1080 60fps. So I didn't expect better results before using it.
> 
Point is not just having 1080p run at 60 FPS or frame drops, but having a bearable amount of CPU usage aswell. Currently normal 720p @ 30 FPS videos cause app to use 15-20% of the CPU along separate GPU work.

In either case, changes are required to Flutter engine itself not a package/plugin.

**Comment by sudongg** [comment]
@alexmercerind 
I don't know why dart_vlc has such a high usage, but I just tested playing the same video (1080 60 fps), it should be a reference
I use LibVLC + LibVLCVideoSetCallbacks + SkiaSharp to play cpu 10% gpu 10%.
uwp +CompositeSurface+ MediaFoundation+ ffmpeg hardware decoding: cpu 2% ~ 5% gpu 10% ~ 15%
QT + openGL+ LibMpv default setting: cpu 10% gpu 10% ~ 30%
But dart_vlc cpu 15% gpu 22% .

**Comment by unoravi** [comment]
Is there any update on Video Play Support for Windows ?

**Comment by azchohfi** [comment]
This work is still depending on https://github.com/flutter/packages/pull/476. Yet, there is another workstream towards enabling this from a PaltformView perspective, which is also evolving. No guarantees on dates, unfortunately.

**Comment by alexmercerind** [comment]
> @alexmercerind I don't know why dart_vlc has such a high usage, but I just tested playing the same video (1080 60 fps), it should be a reference I use LibVLC + LibVLCVideoSetCallbacks + SkiaSharp to play cpu 10% gpu 10%. uwp +CompositeSurface+ MediaFoundation+ ffmpeg hardware decoding: cpu 2% ~ 5% gpu 10% ~ 15% QT + openGL+ LibMpv default setting: cpu 10% gpu 10% ~ 30% But dart_vlc cpu 15% gpu 22% .


So I've made a new video widget (try running example app in [native-video-impl](https://github.com/alexmercerind/dart_vlc/tree/native-video-impl) branch, Windows-only).
In general CPU usage should be 5-10%.

[1080p on Ryzen 3 2200U (outdated ZEN I CPU)]

https://user-images.githubusercontent.com/28951144/162072669-c062ed2a-bbff-4f41-a0c0-8fd4cd64a385.mp4

It's powered entirely by [flutter_native_view](https://github.com/alexmercerind/flutter_native_view) & doesn't use Flutter's texture APIs at-all. The VLC video output is natively embedded.
Currently, there are bugs that I'm aware of. Everything in that tree is highly-experimental. 

But still, 5-10% is WAY better than previous 40-50% pixel buffer impl.

**Comment by lucasdidur** [comment]
They got merged, and now, what is the next steps?

**Comment by Finni123** [comment]
Any update on this? (https://github.com/flutter/packages/pull/476)[https://github.com/flutter/packages/pull/476] got merged, what is the problem now?

**Comment by JuEnPeHa** [comment]
I'm joining to this thread, I'm working with an Camera RTSP project, I need to show the live video in Windows/MacOS.

Still not clean, smooth and resource-less solution?, I'm using dart_vlc for the prototype.

**Comment by lzzy12** [comment]
> I'm joining to this thread, I'm working with an Camera RTSP project, I need to show the live video in Windows/MacOS.
> 
> Still not clean, smooth and resource-less solution?, I'm using dart_vlc for the prototype.

I can't understand how can a platform be considered as "stable" when it doesn't even have a stable and official video player support

**Comment by wang-bin** [comment]
A new choice https://github.com/wang-bin/fvp

**Comment by amir14a** [comment]
Any updates?

**Comment by dbsxdbsx** [comment]
I hope windows could be support ASAP.

**Comment by alexmercerind** [comment]
> I hope windows could be supported ASAP.

I've created [package:media_kit](https://github.com/alexmercerind/media_kit) & it provides hardware accelerated, performant video playback with wide codec support. 
You can visit project page to see the initial guide, implementation details, goals & class diagrams.

4K video playback on entry-level AMD Ryzen 3 2200U processor with Radeon Vega 3 Mobile Graphics:

https://user-images.githubusercontent.com/28951144/208765832-416313c9-97d4-44d0-a902-e577f3c4f3f6.mp4

Though my machine isn't powerful enough to play 4K without frame drops, 1080p 60 FPS is working flawlessly with minimal CPU/GPU load. And I feel it's good & very stable.

Currently, Windows is implemented. Linux support is coming in upcoming week, macOS sometime soon. Support for mobile platforms can also be added. I will make first stable release once at-least two platforms are supported. However, you can use it for Windows for now. It's working really well with high resolution or higher FPS videos.


**NOTE:** This has been made possible after large contributions made by other contributors to Flutter itself.

<hr>

I'm the author of  [package:dart_vlc](https://github.com/alexmercerind/dart_vlc) & a number of improvements have been made in Flutter & other things.
This project aims to bring that to reality. However, I myself can't make any promises for the work that's currently undone.

Thanks! Happy new year!

**Comment by dbsxdbsx** [comment]
@alexmercerind, thanks for your suggestion. Also, today I found [meedu_videoplayer](https://github.com/zezo357/flutter_meedu_videoplayer), which I think is the best one for cross-platform at present.

**Comment by alexmercerind** [comment]
@dbsxdbsx

> @alexmercerind, thanks for your suggestion. Also, today I found [meedu_videoplayer](https://github.com/zezo357/flutter_meedu_videoplayer), which I think is the best one for cross-platform at present.

It's just a "wrapper" which also uses [package:dart_vlc](https://github.com/alexmercerind/dart_vlc) (which I authored) internally. As I said in comment(s) above, it uses pixel buffer & is [not hardware accelerated](https://github.com/alexmercerind/dart_vlc/issues/345) (will drain battery, cause redundant CPU load & forget 1080p or higher). _It means every frame is copied back to system RAM (from GPU buffers) for rendering. Here, libVLC also seems to cost some substantial computation during encoding the pixel buffer in the specified format._

Since this thread is about Windows support, **the newer implementation in [package:media_kit](https://github.com/alexmercerind/media_kit) is GPU backed i.e. H/W accelerated** (& a step forward). It does not use libVLC, but libmpv. It uses [ANGLE](https://github.com/google/angle) (also used in Flutter) to interop between [libmpv's OpenGL API](https://github.com/mpv-player/mpv/blob/master/libmpv/render.h) & [Flutter's Direct3D texture API](https://github.com/flutter/engine/pull/26840). This setup renders the video output to the [`Texture`](https://api.flutter.dev/flutter/widgets/Texture-class.html) widget. I've also added a fallback S/W rendering, just in case. Furthermore, the size of package is also smaller (will be more smaller once we strip down `mpv-2.dll` further after removing "unused" sub-libraries statically linked in it).

Since, [video_player](https://pub.dev/packages/video_player) is federated, an implementation can be provided in a separate package to make it support Windows.

**Comment by leeyisoft** [comment]
> > 我希望能尽快支持windows。
> 
> 我创建了[package:media_kit](https://github.com/alexmercerind/media_kit) & 它提供了硬件加速、高性能视频播放和广泛的编解码器支持。 您可以访问项目页面以查看初始使用文档、实现细节、目标和类图。
> 
> 在配备 Radeon Vega 3 移动显卡的入门级 AMD Ryzen 3 2200U 处理器上播放 4K 视频：
> 
>  media_kit.mp4 
> 虽然我的机器不够强大，无法在不掉帧的情况下播放 4K，但在我的慢速机器上，任何低于或等于 1080p 60 FPS 的东西都可以在最小的 GPU 负载（几乎没有任何 CPU 负载）下完美运行。我觉得它很好而且很稳定。
> 
> 目前，Windows 已经实现（根据我的经验，它是稳定的）。Linux 支持将在下周到来，macOS 也即将到来。还可以添加对移动平台的支持。一旦至少支持两个平台，我将首先“发布”。但是，您现在可以将它用于 Windows。它在处理高分辨率或更高 FPS 视频时效果非常好。
> 
> **笔记：**
> 
> 在大型 Flutter 收到来自全球各种开发人员的大量贡献/评论后，这才成为可能。 无论如何，这不是我唯一的努力。
> 
> 我也是 package:dart_vlc 的作者，并且在 Flutter 和其他方面进行了许多改进。 该项目旨在将其变为现实。但是，我本人无法对目前未完成的工作做出任何承诺。
> 
> 谢谢！新年快乐！

Can I replace video_player with media_kit or dart_vlc?

I'm still not sure how different media_kit is from dart_vlc, I'm making a chat APP(https://github.com/imboy-pub/imboy-flutter), and since video_player doesn't support macOS, After the release of your two excellent projects, how can I introduce them into my own projects?

**Comment by leeyisoft** [comment]
> > I hope windows could be supported ASAP.
> 
> I've created [package:media_kit](https://github.com/alexmercerind/media_kit) & it provides hardware accelerated, performant video playback with wide codec support. You can visit project page to see the initial usage docs, implementation details, goals & class diagrams.
> 
> 4K video playback on entry-level AMD Ryzen 3 2200U processor with Radeon Vega 3 Mobile Graphics:
> 
>  media_kit.mp4 
> Though my machine isn't powerful enough to play 4K without frame drops, anything below or equal to 1080p 60 FPS is working flawlessly with minimal GPU load (& barely any CPU load) on my slow machine. I feel it's good & very stable.
> 
> Currently, Windows is implemented (& stable from my experience). Linux support is coming in upcoming week, macOS sometime soon. Support for mobile platforms can also be added. I will make first "release" once at-least two platforms are supported. However, you can use it for Windows for now. It's working really well with high resolution or higher FPS videos.
> 
> **NOTE:**
> 
> This has been made possible after large Flutter received large contributions/reviews from various developers worldwide. It's not my sole effort by any means.
> 
> I'm the author of package:dart_vlc aswell & a number of improvements have been made in Flutter & other things. This project aims to bring that to reality. However, I myself can't make any promises for the work that's currently undone.
> 
> Thanks! Happy new year!

Can I replace video_player with media_kit or dart_vlc?

I'm still not sure how different media_kit is from dart_vlc, I'm making a chat APP(https://github.com/imboy-pub/imboy-flutter), and since video_player doesn't support macOS, Then I found these two excellent projects of yours, how can I introduce them into my projects?

**Comment by wang-bin** [comment]
I've published [fvp](https://pub.dev/packages/fvp) on pub.dev. Now it's a video_player plugin supports all mobile and desktop platforms. Just adding 2 lines in your existing video_player code. Decoder and renderer are gpu accelerated, cpu load should be lower than mpv.

**Comment by alexmercerind** [comment]
> I've created [package:media_kit](https://github.com/alexmercerind/media_kit) & it provides hardware-accelerated, performant video playback with wide codec support.

[package:media_kit](https://github.com/alexmercerind/media_kit) is now 1.0.0. It supports all platforms: Windows, GNU/Linux, macOS, iOS, Android & Web.

Aside from good performance & stability... There is support for highly-customizable pre-built video controls (Material: available; Cupertino: soon), playlists, audio device selection, video/audio/subtitle track selection etc. A [large number of unit-tests have been written](https://github.com/alexmercerind/media_kit/blob/main/media_kit/test/src/player/player_test.dart) in order to focus on stability, avoid regressions & maintain parity between platforms.

The project is split into packages for modularity & shares same API for both video & audio playback (both are supported).

I will try my best to address any issues that the community reports :)

Alternative platforms-specific backends may also be implemented, in-case required.

<table>
  <tr>
    <td>
      Android
    </td>
    <td>
      iOS
    </td>
  </tr>
  <tr>
    <td>
      <img src="https://github.com/alexmercerind/media_kit/assets/28951144/cf93a1fd-e1d8-4d1c-8bd5-cc393cef1ce9" height="400" alt="Android"></img>
      <img src="https://github.com/alexmercerind/media_kit/assets/28951144/aea1f480-51e2-452a-b53c-c0e27f71f0d8" height="400" alt="Android"></img>
    </td>
    <td>
      <img src="https://github.com/alexmercerind/media_kit/assets/28951144/e8ce64cb-1ea9-4a3e-bc9c-db620abf88c9" height="400" alt="iOS"></img>
      <img src="https://github.com/alexmercerind/media_kit/assets/28951144/d7159df2-1df1-46d3-84f8-238e2a66bfbc" height="400" alt="iOS"></img>
    </td>
  </tr>
  <tr>
    <td>
      macOS
    </td>
    <td>
      Windows
    </td>
  </tr>
  <tr>
    <td>
      <img src="https://github.com/alexmercerind/media_kit/assets/28951144/fca8dbbf-4262-431f-a04a-f3aa6afb2911" height="200" alt="macOS"></img>
    </td>
    <td>
      <img src="https://github.com/alexmercerind/media_kit/assets/28951144/742b0016-da58-42de-9880-ecaa0604c2b2" height="200" alt="Windows"></img>
    </td>
  </tr>
  <tr>
    <td>
      GNU/Linux
    </td>
    <td>
      Web
    </td>
  </tr>
  <tr>
    <td>
      <img src="https://github.com/alexmercerind/media_kit/assets/28951144/8cd63750-6746-4c75-bc4e-cca5e4c61890" height="200" alt="GNU/Linux"></img>
    </td>
    <td>
      <img src="https://github.com/alexmercerind/media_kit/assets/28951144/feb9fdf2-095f-43db-96af-f7782985238d" height="200" alt="Web"></img>
    </td>
</table>

**Comment by stuartmorgan** [comment]
It would be helpful if everyone could read https://github.com/flutter/flutter/issues/37673#issuecomment-637010916 and https://github.com/flutter/flutter/wiki/Issue-hygiene#comments-providing-workarounds

Someone is welcome to make a forum post about alternate video player packages and link to it here, and then there can be as much discussion as people want there about other packages.

**Comment by FMorschel** [comment]
Any updates here? What is missing so this can move forward?

**Comment by stuartmorgan** [comment]
> What is missing so this can move forward?

See https://github.com/flutter/flutter/issues/37673#issuecomment-918599475 for previous discussion of potential work.

**Comment by FMorschel** [comment]
The [mentioned PR](https://github.com/flutter/engine/pull/24705) is closed, what can we expect?

**Comment by azchohfi** [comment]
I've done a little bit more investigation, and it seems like flutter still doesn't support Direct Composition. I thought it did. What it does support is to render a ID3D11Texture2D, which we can get from Media Foundation, but it is not optimal, because it has to be frame by frame.
To have this as efficient as possible, without copying ID3D11Texture2D, we would need #76789, or some other mechanism to pass down the actual swap chain from the IMFMediaEngine.

I will try to investigate this further, but I did get this working:
![image](https://github.com/flutter/flutter/assets/527713/febd31ba-ad04-4eea-82dd-387f0050a6a4)
It might seem like it is working, and it kind of is, but the refresh(rate) of the image is happening only at random events that I could call my Update method.
We could add a 60fps times to get the latest, but I don't think that is a good idea.

**Comment by azchohfi** [comment]
Actually... I kind of mostly got it working, and with decent perf \o/
The way to go is to have a background thread pumping the textures from Media Engine to the ID3D11Texture2D. It is mostly happening on the GPU, so the only thing we need that thread for is to ask for the transfer, and properly keep it in v-sync (which I haven't done yet). I've fixed some memory issues and it is now looking quite nice! I'm almost ready to open a PR. I'm checking with some colleagues on the right Media Foundation pattern for this, to make sure I'm not doing anything that I shouldn't, but it is looking good.

**Comment by gtgalone** [comment]
https://pub.dev/packages/video_player_win

this package solved windows support

**Comment by syleishere** [comment]
> https://pub.dev/packages/video_player_win
> 
> this package solved windows support

Thought everyone moved on to mediakit

**Comment by gtgalone** [comment]
@syleishere wow! mediakit is amazing! I will use it. Thanks!

**Comment by cjng96** [comment]
😁

2024년 10월 9일 (수) 오전 6:05, syle ***@***.***>님이 작성:

**Comment by iampopal** [comment]
When this feature is expacted to add to windows?
